---
title: "Heatmap: All stages modENCODE ChIP-seq overlapping promoters of cell sorted intestine RNA-seq"
author: "David C. King"
date: "03/27/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache=TRUE)
library(tidyverse)
library(dplyr)
library(GenomicRanges)
library(ComplexHeatmap)
library(RMariaDB)
library(R.cache)
library(magrittr)
library(stringr)

# for distances between rows/columns in a binary matrix
cosine.dist = function(df) { 
  M = as.matrix(df)
  sim = M / sqrt(rowSums(M*M))
  sim = sim %*% t(sim)
  as.dist(1-sim)
}
source('DBPeakOverlaps.R', echo = T) # sources DB_connect.R
```

```{r intestine present genes}
intestinePresentDB = dbReadTableCached("intestine_present", "NishimuraLab")
cat(nrow(intestinePresentDB), "genes enriched or equal in any stage")
head(intestinePresentDB,10)
```

```{r }
#DBPeakOverlaps %<>% column_to_rownames("name")
cat(nrow(DBPeakOverlaps), "genes with ChIP-seq overlap information")
DBPeakOverlaps[1:10,1:10]
```

```{r merge overlaps with intestine-present}
DBPeakOverlapsIntestinePresent = intestinePresentDB %>% inner_join(DBPeakOverlaps, by = "name")
enrichedmentPattern = DBPeakOverlapsIntestinePresent %>% select(name, WBID, embryo, L1, L3)
DBPeakOverlapsIntestinePresent %<>% column_to_rownames("name") %>% select(-WBID, -embryo, -L1, -L3)


sum(!is.na(DBPeakOverlapsIntestinePresent))
naOrZero = function(x) { is.na(x) | (x == 0)}
mx = ifelse(naOrZero(DBPeakOverlapsIntestinePresent), 0, 1)

```


```{r count zero fraction}

fractionZero = sum(mx == 0) / length(mx)
cat("fraction of zero cells:", fractionZero)
```


```{r make-into-binary}


rs = rowSums(mx)
cs = colSums(mx)
cat("The following genes are bound in over 90% of the ChIP-seq experiments:")
rownames(mx)[rs > .9*ncol(mx)]
```

```{r euclidean distance}
system.time({eu.mx.row = dist(mx)})
  #  user  system elapsed 
  # 16.157   0.057  16.228 
system.time({eu.mx.col = dist(t(mx))})
  #  user  system elapsed 
  # 2.311   0.010   2.322 
system.time({hc.mx.rows = hclust(eu.mx.row, method = "average")})
  #  user  system elapsed 
  # 0.536   0.037   0.577 
system.time({hc.mx.cols = hclust(eu.mx.col, method = "average")})
  #  user  system elapsed 
  # 0.005   0.001   0.006
```

```{r cosine distance}
system.time({cos.mx.row = cosine.dist(mx)})
  #    user  system elapsed 
  # 7.337   0.214   7.590 
system.time({cos.mx.col = cosine.dist(t(mx))})
  #  user  system elapsed 
  # 0.845   0.010   0.860 

system.time({hc.mx.cosine.rows = hclust(cos.mx.row, method = "average")})
  # user  system elapsed 
  # 0.542   0.041   0.587 
system.time({hc.mx.cosine.cols = hclust(cos.mx.col, method = "average")})
  #  user  system elapsed 
  # 0.005   0.000   0.006 
```
```{r scree}
plot(1:length(hc.mx.rows$height),rev(hc.mx.rows$height),  xlim=c(0,50), main="scree for euclidean rows")
plot(1:length(hc.mx.cols$height),rev(hc.mx.cols$height),  xlim=c(0,50), main="scree for euclidean columns")

plot(1:length(hc.mx.cosine.rows$height),rev(hc.mx.cosine.rows$height),  xlim=c(0,50), main="scree for cosine rows")
plot(1:length(hc.mx.cosine.cols$height),rev(hc.mx.cosine.cols$height),  xlim=c(0,50), main="scree for cosine columns")
```

```{r}

hmColors = c('#FFFFCC',  '#68001C')

hm.euclidean = Heatmap(mx, 
        cluster_rows = hc.mx.rows, 
        cluster_columns = hc.mx.cols, 
        heatmap_legend_param = list(title = "euclidean\nTF bound/unbound"),
        show_heatmap_legend = FALSE,
        show_row_names = FALSE, 
        show_column_names = T, 
        show_row_dend = F,
        show_column_dend = T,
        col = hmColors,
        row_split = 4,
        column_split = 5,
        column_names_side = "top",
        use_raster = T)

hm.cosine = Heatmap(mx, 
        cluster_rows = hc.mx.cosine.rows, 
        cluster_columns = hc.mx.cosine.cols, 
        heatmap_legend_param = list(title = "cosine\nTF bound/unbound"),
        show_heatmap_legend = FALSE,
        show_row_names = FALSE, 
        show_column_names = T, 
        show_row_dend = F,
        show_column_dend = T,
        col = hmColors,
        column_split = 5,
        column_names_side = "top",
        use_raster = T)


system.time(print(hm.euclidean))
 #   user  system elapsed 
 # 11.654   2.262  13.958 
system.time(print(hm.cosine)) 
 #   user  system elapsed 
 # 12.657   2.282  14.993 


pdf(file=format(Sys.time(), "plots/%Y-%m-%d-all-heatmap-euc.pdf"), 
    width=60, height=20)
hm.euclidean
dev.off()

pdf(file=format(Sys.time(), "plots/%Y-%m-%d-all-heatmap-cos.pdf"), 
    width=60, height=20)
hm.cosine
dev.off()
```
```{r bin}
DBPeakOverlapsBin = ifelse(DBPeakOverlaps > 0, 1,0)
mx.bin = as.matrix(DBPeakOverlapsBin)
rs = rowSums(DBPeakOverlapsBin)
cs = colSums(DBPeakOverlapsBin)
cat("The following genes are bound in over 90% of the ChIP-seq experiments:")
rownames(DBPeakOverlapsBin)[rs > .9*ncol(DBPeakOverlapsBin)]
```

```{r euclidean distance}
system.time({eu.mx.row = dist(mx.bin)})
  #  user  system elapsed 
  # 218.859   0.807 219.832
system.time({eu.mx.col = dist(t(mx.bin))})
  #  user  system elapsed 
  # 8.565   0.036   8.601 
system.time({hc.mx.rows = hclust(eu.mx.row)})
  #  user  system elapsed 
  # 6.414   2.053   9.672 
system.time({hc.mx.cols = hclust(eu.mx.col)})
  #  user  system elapsed 
  # 0.005   0.004   0.015 
```

```{r cosine distance}
system.time({cos.mx.row = cosine.dist(mx.bin)})
  #    user  system elapsed 
  # 217.185   2.265 219.539 
system.time({cos.mx.col = cosine.dist(t(mx.bin))})
  #  user  system elapsed 
  # 2.730   0.014   2.746 

system.time({hc.mx.cosine.rows = hclust(cos.mx.row)})
  # user  system elapsed 
  # 5.656   0.382   6.046 
system.time({hc.mx.cosine.cols = hclust(cos.mx.col)})
  #  user  system elapsed 
  # 0.006   0.001   0.006 
```
```{r scree}
plot(1:length(hc.mx.rows$height),rev(hc.mx.rows$height),  xlim=c(0,50), main="scree for euclidean rows")
plot(1:length(hc.mx.cols$height),rev(hc.mx.cols$height),  xlim=c(0,50), main="scree for euclidean columns")

plot(1:length(hc.mx.cosine.rows$height),rev(hc.mx.cosine.rows$height),  xlim=c(0,50), main="scree for cosine rows")
plot(1:length(hc.mx.cosine.cols$height),rev(hc.mx.cosine.cols$height),  xlim=c(0,50), main="scree for cosine columns")
```

```{r}
countTable = table(mx.bin)
nlevels = length(countTable)

hmColors = c("#FFFFCC",   "#BE9180", "#B37F74", "#A86D67", "#9D5B5A", "#93484E", "#883641",  "#7D2435", "#721228", "#68001C")
# adjust levels of palette
hmColors = c('#FFFFCC', colorRampPalette(c("#A86D67", '#68001C'))(nlevels-1))

hm.euclidean = Heatmap(mx.bin, 
        cluster_rows = hc.mx.rows, 
        cluster_columns = hc.mx.cols, 
        heatmap_legend_param = list(title = "# times a TF is bound in gene promoter"),
        show_row_names = FALSE, 
        show_column_names = FALSE, 
        show_row_dend = T,
        show_column_dend = T,
        col = hmColors,
        use_raster = T)

hm.cosine = Heatmap(mx.bin, 
        cluster_rows = hc.mx.cosine.rows, 
        cluster_columns = hc.mx.cosine.cols, 
        heatmap_legend_param = list(title = "# times a TF is bound in gene promoter"),
        show_row_names = FALSE, 
        show_column_names = FALSE, 
        show_row_dend = T,
        show_column_dend = T,
        col = hmColors,
        use_raster = T)


system.time(print(hm.euclidean))
system.time(print(hm.cosine)) 
 #   user  system elapsed 
 # 26.277   2.839  29.193 
 #   user  system elapsed 
 # 33.472   2.916  36.446 
```

