---
title: "Intestine promoter binding"
author: "David C. King"
date: "11/11/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(dplyr)
library(GenomicRanges)
#library(pheatmap)
library(ComplexHeatmap)
library(RMariaDB)

# distance.cosine = function(u,v) {
#   1 - ((u%*%v) / (norm(u,"2")*norm(v,"2")))
# }
cosine_dist = function(df) { # much faster
  M = as.matrix(df)
  sim = M / sqrt(rowSums(M*M))
  sim = sim %*% t(sim)
  as.dist(1-sim)
}
```

See https://stats.stackexchange.com/questions/31565/compute-a-cosine-dissimilarity-matrix-in-r/367216#367216 for the efficient computation of the cosine dissimilarity.

## Goal

This analysis is looking for TFs that may explain intestine expression in the absence of ELT-2.

## Promoters

Genes which fall under different classes of intestine enrichment in Williams et al. 2023
are assigned promoters gene 5' -1Kb/200bp. 

The dataset has each promoter in the genome annotated with intestine expression DESeq information, as well as ELT-2 binding status gotten from Kudron et al. 2018.
```{r wtf3}
onishDATA <- dbConnect(
     drv = RMariaDB::MariaDB(), 
     username = "root",
     password = "candleWax1!", 
     host = "129.82.125.11", 
     port = 3307, dbname = "NishimuraLab"
)
wtf = dbReadTable(onishDATA, "WTF3")
dbDisconnect(onishDATA)
```


```{r promoters}
promoters_df = readRDS('DATA/promoter_data_2022_12_06.rds')
promoters_df %>% head()
rownames(promoters_df) <- promoters_df$name
geneNamesToWBID = promoters_df %>% select(name,WBGeneID)
```

### Promoters by intestine enrichment status



```{r enrichment_depletion}
intestine_present = promoters_df %>% filter(embryo_int_exp %in% c("enriched", "equal") | 
                                        L1_int_exp %in% c("enriched", "equal") | 
                                        L3_int_exp %in% c("enriched", "equal"))


promoters_gr = makeGRangesFromDataFrame(intestine_present, keep.extra.columns = T, seqinfo = Seqinfo(genome="ce11"))
```

```{r other-TFs}

mapping = read.table("TFNameMapping.df",header = T)

# label duplicates
mapping$name_stage = paste(mapping$name, mapping$stage, sep="_")
with_dups = mapping %>% 
  filter(name_stage %in% mapping[which(duplicated(mapping$name_stage)), 'name_stage']) %>% 
   group_by(name_stage) %>% mutate(id = row_number()) 
with_dups$name_stage = paste(with_dups$name_stage,name_stage, sep=".")

rownames(mapping) <- mapping$path
mapping[with_dups$path, 'name_stage'] = with_dups$name_stage

TF.files = mapping %>% pull(path)
names(TF.files) = mapping  %>% pull(name_stage)
TFs = geneNamesToWBID[names(TF.files),]


all.grs= lapply(TF.files, 
  function(fname) {
    
    data=read.table(fname)
    gr = makeGRangesFromDataFrame(data, keep.extra.columns = T, 
                                  seqnames.field = "V1",
                                  start.field = "V2",
                                  end.field = "V3", seqinfo=Seqinfo(genome="ce11"))
    return(gr)
  })
```

```{r}


for(TFname in names(all.grs))
{
  
  olap=countOverlaps(promoters_gr,all.grs[[TFname]])
  
  df = mcols(promoters_gr)# pull dataframe from GRanges metadata
  df[[TFname]] = olap
  
  mcols(promoters_gr) <- df # reassign dataframe to GRanges
   
}
```

```{r}

data = as.data.frame(mcols(promoters_gr))
data =data[,42:ncol(data)]
moddata= data[rowSums(data)!=0,]
colSums(as.data.frame(data))
```

```{r}
mtx_x <- matrix(unlist(moddata), ncol = length(moddata), byrow = TRUE)
```

```{r scaled-hclust-euclidean eval=FALSE}
heatmap(mtx_x,scale="none")
```

```{r}
mx = as.matrix(data)
mx.filt = mx[rowSums(mx) >0, ]
table(rowSums(mx.filt))
cat("fraction of zero values: ")
sum(mx.filt==0)/length(mx.filt)
```
# No filtering

```{r presence-absence-of-binding}

mx.ind = ifelse(mx.filt > 0, 1, 0)  
rs = rowSums(mx.ind)


# have to remove any full 0 rows/columns for clustering with cos
mx.ind = mx.ind[rs > 2,]
cs = colSums(mx.ind)
mx.ind = mx.ind[,cs > 0]
stopifnot(sum(rowSums(mx.ind) == 0)==0)
stopifnot(sum(colSums(mx.ind) == 0)==0)
```

```{r apply-cos-dist}
dist_cos_rows = cosine_dist(mx.ind)
dist_cos_cols = cosine_dist(t(mx.ind))
```

```{r hclust-for-screeplot, eval=TRUE}
hc.cos = hclust(dist_cos_rows, method="average")
plot(1:length(hc.cos$height),rev(hc.cos$height),  xlim=c(0,50), main="scree for cos.")
#clusterboot(dist_cos_rows, B=100, bootmethod="boot",clustermethod=disthclustCBI, method="average", k=2)
#hc.cos.k = cutree(hc.cos, 10)
hc.cos = hclust(dist_cos_cols, method="average")
plot(1:length(hc.cos$height),rev(hc.cos$height),  xlim=c(0,100), main="scree for cos. on cols")
```

```{r pheatmap-add-int.enriched}
rownames(intestine_enriched) <- sub("-",".", intestine_enriched$name)
int.enriched = data.frame(intestine=factor( ifelse(intestine_enriched[,'enriched'],'yes','no'))) 

rownames(int.enriched) = sub("-",".", intestine_enriched$name)
```

```{r pheatmap-plot}
ph=pheatmap(mx.ind, clustering_distance_rows = dist_cos_rows,
         clustering_distance_cols = dist_cos_cols,
         cutree_rows = 10,
         cutree_cols = 14,
         annotation_col = int.enriched,
         show_rownames = F,
         clustering_method = "average",
         color = c('#FFFFCC', '#68001C'), # to match heatmap() colors with 0/1
         main=sprintf("%s TF presence/absence (filter: TF must bind > 2 promoters)", STAGE), 
         legend=FALSE
         )
pdf(sprintf("hc.%s.noFilter.pdf", STAGE), width=11.5,height=8)
ph
dev.off()
```

```{r ComplexHeatmap}
Heatmap(mx.ind, 
        name = "TF bound",
        col = c('#FFFFCC', '#68001C'),
        width = 8,
        # row settings
        clustering_distance_rows = dist_cos_rows, 
        clustering_method_rows = "average",
        show_row_names = F,
        show_row_dend = F,
        row_title = "Intestine Present (Embryo)",
        row_title_rot = 90,
        left_annotation = rowAnnotation(
          WTF = rownames(mx.ind) %in% wtf$name,           
          col = list(WTF = c("TRUE" = "black", "FALSE" = "white"))
        ),
        
        # column settings
        clustering_distance_columns = dist_cos_cols,
        clustering_method_columns = "average",
        column_labels = intestine_enriched$name,
        column_title = "ChIP-seq merged peaks bound to Promoters (whole worm)",
        column_names_rot = 45,
        bottom_annotation = columnAnnotation(
          intestine_present = intestine_enriched$enriched,
          col = list(intestine_present = c("TRUE" = "black", "FALSE" = "#F8F8F8"))
        )
        )
        

```
## Adjacency matrix
```{r adjacency}
mx.tf = mx
colnames(mx.tf) <- intestine_enriched$name
adj = mx.tf[rownames(mx.tf) %in% wtf$name,int]
adj.present = adj[rowSums(adj)>0,]
adj.present = adj.present[,colSums(adj.present)>0]
adj.present.ind = ifelse(adj.present > 0, 1, 0)

Heatmap(adj.present.ind, 
        clustering_distance_rows = cosine_dist(adj.present.ind),
        clustering_distance_columns = cosine_dist(t(adj.present.ind)), 
        clustering_method_rows = "average",col = c('#FFFFCC', '#68001C'))


network <- graph_from_adjacency_matrix(t(full.mx), mode="directed")
node.size= c(10,10,10)*1
plot(network, layout=layout_with_fr, rescale=F,axes=TRUE,ylim=c(-5,6),xlim=c(0,10))
```

## Filtered

```{r filtered, eval=FALSE}
mx.ind = ifelse(mx.filt > 0, 1, 0)  
rs = rowSums(mx.ind)
mx.ind = mx.ind[rs > 2 & rs < 20,]
cs = colSums(mx.ind)
mx.ind = mx.ind[,cs > 0]
stopifnot(sum(rowSums(mx.ind) == 0)==0)
stopifnot(sum(colSums(mx.ind) == 0)==0)
```

```{r apply-cos-dist-filtered, eval=FALSE}
dist_cos_rows = cosine_dist(mx.ind)
dist_cos_cols = cosine_dist(t(mx.ind))
```

```{r hclust-for-screeplot-filtered, eval=FALSE}
hc.cos.rows = hclust(dist_cos_rows, method="average")
plot(1:length(hc.cos.rows$height),rev(hc.cos.rows$height),  xlim=c(0,100), main="scree for cos. on rows")
hc.cos.cols = hclust(dist_cos_cols, method="average")
plot(1:length(hc.cos.cols$height),rev(hc.cos.cols$height),  xlim=c(0,100), main="scree for cos. on cols")
```

```{r pheatmap-filtered, eval=FALSE}
ph=pheatmap(mx.ind, clustering_distance_rows = dist_cos_rows,
         clustering_distance_cols = dist_cos_cols,
         cutree_rows = 8, 
         cutree_cols = 21, 
         annotation_col = int.enriched,
         show_rownames = F,
         clustering_method = "average",
         color = c('#FFFFCC', '#68001C'), # to match heatmap() colors with 0/1
         main=sprintf("%s TF presence/absence ", STAGE), 
         legend=FALSE)
pdf(file=sprintf("hc.%s.noHOT.pdf", STAGE),width=11.5,height=6) 
ph
dev.off()
```

```{r cluster-stability, eval=FALSE}

clusterboot(dist_cos_rows, B=1000, bootmethod="boot",clustermethod=disthclustCBI, method="average", k=13,seed=110)

hc.cos.rows.k = cutree(hc.cos.rows, k=13)
table(hc.cos.rows.k)
```

```{r try-kmeans, eval=FALSE}
clusterboot(mx.ind, B=100, bootmethod="boot",clustermethod=kmeansCBI,k=2)
```

```{r pheatmap-filtered-kmeans, eval=FALSE}
pheatmap(mx.ind, kmeans_k=2,
         clustering_distance_cols = dist_cos_cols,
         clustering_method = "average",
         color = c('#FFFFCC', '#68001C'), # to match heatmap() colors with 0/1
         main="TF presence/absence ", 
         legend=FALSE
         #, file="hc.L1.noHOT.pdf" # uncomment to write to a file instead of plot
         )
```
